radiometric_normalization/pif.py

```diff
@@ -72,3 +81,156 @@ def _filter_zero_alpha_pifs(reference_gimage, candidate_gimage):
+
+        PIF_band = _PCA_fit_and_filter_single_band(
+            candidate_gimage.bands[band_no].flatten(),
+            reference_gimage.bands[band_no].flatten(),
+            valid_pixels_list, array_shape, lim)
+        PIF_all = numpy.logical_and(PIF_band, PIF_all)
+
+    if logging.getLogger().getEffectiveLevel() <= logging.INFO:
+        for band_no in xrange(len(candidate_gimage.bands)):
+            cand_combo_filtered = candidate_gimage.bands[band_no][
+                numpy.nonzero(PIF_all != 0)].flatten()
+            ref_combo_filtered = reference_gimage.bands[band_no][
+                numpy.nonzero(PIF_all != 0)].flatten()
+
+            logging.info('PCA Info: Filtered combo corrcoef (band {}) ='
+                         ' {}'.format(band_no, numpy.corrcoef(
+                            cand_combo_filtered, ref_combo_filtered)[0, 1]))
+
+    no_valid_pixels = len(numpy.nonzero(PIF_all.flatten() != 0)[0])
+    no_total_pixels = candidate_gimage.bands[0].size
+    valid_percent = 100.0 * no_valid_pixels / no_total_pixels
+    logging.info('PCA Info: Found {} final pifs out of {} pixels ({}%)'.format(
+        no_valid_pixels, no_total_pixels, valid_percent))
+
+    return 65535 * PIF_all.astype(numpy.uint16)
+
+
+def _PCA_fit_and_filter_single_band(candidate_band, reference_band,
+                                    valid_pixels_list, array_shape, lim):
+    ''' Performs PCA analysis, on the valid pixels and filters according
+    to the distance from the principle eigenvector, for a single band.
+    '''
+
+    cand_valid = candidate_band[valid_pixels_list]
+    ref_valid = reference_band[valid_pixels_list]
+
+    if logging.getLogger().getEffectiveLevel() <= logging.INFO:
+        logging.info('PCA Info: Original corrcoef = {}'.format(
+            numpy.corrcoef(cand_valid, ref_valid)[0, 1]))
+
+    fitted_pca = _PCA_fit_single_band(cand_valid, ref_valid)
+
+    if logging.getLogger().getEffectiveLevel() <= logging.INFO:
+        for component_no in xrange(len(fitted_pca.components_)):
+            logging.info('PCA Info: Component {} is {} with an explained '
+                         'variance of {}'.format(
+                            component_no,
+                            fitted_pca.components_[component_no],
+                            fitted_pca.explained_variance_ratio_[component_no])
+                         )
+
+    def split_into_batches(array, no_per_batch):
+        return [array[i:i + no_per_batch]
+                for i in xrange(0, len(array), no_per_batch)]
+
+    no_per_batch = 10000
+    cand_valid_batches = split_into_batches(cand_valid, no_per_batch)
+    ref_valid_batches = split_into_batches(ref_valid, no_per_batch)
+
+    logging.info('PCA Info: Dataset split into {} batches of {} '
+                 'each'.format(len(cand_valid_batches),
+                               no_per_batch))
+
+    passed_pixels = [[]]
+    for batch_no in xrange(len(cand_valid_batches)):
```

Comment:
Same goes here, any time these python loops can be avoids, the better. This is a tough task to accomplish whenever applying these ML algos over image space. This might be fine for your purposes, but I think we'll continue to run into situations like there where it's best to push the loops down to C.